cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+non-intuitive, it is best to wrap it within friendlier internal
+extractors.
+
+### Mapping Type Declarations to Rust
+
+When we declare a type like
+
+```lisp
+    (decl MyEnum (enum
+                   (A (x u32) (y u32))
+                   (B)))
+```
+
+ISLE will generate the Rust type definition
+
+```rust
+#[derive(Clone, Debug)]
+pub enum MyEnum {
+    A { x: u32, y: u32, },
+    B,
+}
+```
+
+Note that enum variants with no fields take on the brace-less form,
+while those with fields use the named-struct-field `A { x: ... }`
+form. If all variants are field-less, then the type will additionally
+derive `Copy`, `PartialEq`, and `Eq`.
+
+If the type is declared as extern (`(decl MyEnum extern (enum ...))`)
+then the same definition is assumed to exist. Primitives (`(decl u32
+(primitive u32))`) are assumed to be defined already, and are required
+to be `Copy`.
+
+All imported/extern types are pulled in via `use super::*` at the top
+of the generated code; thus, these types should exist in (or be
+re-exported from) the parent module.
+
+### Symbolic Constants
+
+ISLE allows the user to refer to external constants as follows:
+
+```lisp
+    (extern const $I32 Type)
+```
+
+This allows code to refer to `$I32` whenever a value of type `Type` is
+needed, in either a pattern (LHS) or an expression (RHS). These
+constants are pulled in via the same `use super::*` that imports all
+external types.
+  
+### Exported Interface: Functions and Context Trait
+
+The generated ISLE code provides an interface that is designed to be
+agnostic to the embedding application. This means that ISLE knows
+nothing about, e.g., Cranelift or compiler concepts in
+general. Rather, the generated code provides function entry points
+with well-defined signatures based on the terms, and imports the
+extern constructors and extractors via a trait that the embedder must
+implement.
+
+When a term `T` is defined like
+
+```lisp
+    (decl T (A B C) U)
+```
+
+and has an internal constructor (provided by `rule` bodies), then a
+function with the following signature will be exported from the
+generated code:
+
+```rust
+    pub fn constructor_T<C: Context>(ctx: &mut C, arg0: &A, arg1: &B, arg2: &C) -> Option<U>;
+```
+
+In other words, `constructor_` is prepended, and the function takes
+the expected arguments, along with a "context" (more on this
+below). It returns an `Option<U>` because internal constructors can be
+partial: if no rule's pattern matches, then the constructor
+fails. Note that if a sub-constructor fails, no backtracking occurs;
+rather, the failure propagates all the way up to the entry point.
+
+What is this "context" for? The context argument is used to allow
+external extractors and constructors to access the necessary state of
+the embedding application. (For example, in Cranelift, it might be the
+`LowerCtx` that controls the code-lowering process.)
+
+The context is a trait because we want to decouple the generated code
```

Comment:
This also will make testing/fuzzing much easier, presumably :D

Reply:
Indeed! No concrete plans yet to independently test the generated code but possibly we could come up with an alternate implementation of the input side (the extractors), maybe even inventing the operand tree as we go (carry the `arbitrary::Unstructured` in the context object and then generate things on demand). cc @fitzgen for possible fuzzing opportunity :-)

Reply:
I created #3574 to record this.